### PR TITLE
Private-Events

### PR DESCRIPTION
**Pull request by AbdelrhmanAmin**
In this project, we have built an application named Private Events and that behaves similar to Eventbrite. 
The main goal was to put in practice the main concepts of Associations in rails.

### CODE REVIEW

**Review by MarilenaRoque — COMMENTED** [review]
# Hello, @AbdelrhmanAmin! :rocket:

Congratulations on the effort to complete that project! :clap:

For now, I can't proceed with the review cause the pull request should be made from feature branch to development. Following the git-flow guideline. As requested [here](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59882084/).

# Please, correct that issue and submit it for a new review. 

Happy coding! 💻 

**_Marilena Roque
Full-Stack Developer
Code Reviewer at Microverse Inc.
[Portfolio](https://mariroqueproject.com/) | [Linkedin](https://www.linkedin.com/in/roquemarilena/) 
marilenas@id.uf.br_**

**Review by Temmarie — CHANGES_REQUESTED** [review]
## Changes Required! :recycle:

### Hi @AbdelrhmanAmin  ! :wave: :wave:

#### You have done well so far :clap: However, there are changes required for you to make in order for this project to be approved. Kindly work on them and submit them for another code review.


**Recommended**

- [x] There should be navigation links present at the top of the page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/navigation_links_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)) In a nutshell, make your UI look better, there should be a navigation bar in your application.

- [x] I cannot attend events by other users, so add a button on the event's show page that allows me to do so.
- [x] When I edit and update an event, it registers me as another attendee. This should not be the case here.


__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Comment by mimipeshy** [comment]
### STATUS: CHANGES REQUIRED :red_circle: :recycle: 

Hey there @AbdelrhmanAmin, good job on working on all changes requested by the TSE, keep it up. However, when I run RSpec, you have missed out on testing some methods hence the need to add them before the project is approved;

image

[HINT] You can set up [FactoryBot](https://semaphoreci.com/community/tutorials/working-effectively-with-data-factories-using-factorygirl) which is a helper for writing [factories](https://thoughtbot.com/blog/why-factories) for Ruby tests. This will guide you in writing explicit tests.
- You can refer to this [cheatsheat](https://devhints.io/factory_bot) and this [link](https://www.codewithjason.com/set-factory-bot-fresh-rails-project/) as well for more information.

That's all :smile: , kindly submit another review when done.

_Feel free to reach out to me on slack if you have any questions or something is unclear, I will be glad to help_
Regards
Peris Ndanu

File: app/views/events/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  <tbody>
+    <% @events.coming_events.each do |event| %>
+      <tr>
+        <td><%= event.title %></td>
+        <td><%= event.creator.name %></td>
+        <td><%= event.date %></td>
+        <td><%= link_to 'Show', event, class: 'text-success' %></td>
+        <% if user_signed_in? %>
+          <% if event.creator.id == current_user.id %>
+            <td><%= link_to 'Edit', edit_event_path(event), class: 'text-warning' %></td>
```
Comment (MahmoudBakr23) [inline app/views/events/index.html.erb]: Please refactor the logic in this file and add it to the application_helper.rb

File: app/views/users/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    </div>
+  </div>
+  <div class="row">
+    <div class="col-6">
+      <h4>HOSTING</h4>
+            <ul>
+          <% @user.events.each do |e| %>
+              <% if e.date >= Date.today %>
+              <li><%= e.title %></li><br>
+              <% end %>
```
Comment (MahmoudBakr23) [inline app/views/users/show.html.erb]: Please refactor the logic in this file and add it to the application_helper.rb i.e. you could make a function called dates_manage or whatever you like and using it in the views instead.

File: app/views/events/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  <% if @event.date < Date.today%>
+    <strong class='h5 text-danger'>THIS IS AN OLD EVENT! YOU CAN'T ATTEND IT ANYMORE</strong>
+  <%end%>
+  <br>
+  <br>
+  <%if @event.attendances.length > 0%>
+  <h4 class='text-success'>Attendees</h4>
+    <% @event.attendances.each do |e| %>
+      <%= e.name %> <br>
+      <% if user_signed_in? %>
```
Comment (MahmoudBakr23) [inline app/views/events/show.html.erb]: Please refactor the logic in this file and add it to the application_helper.rb

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
# **Required Changes ♻**

<hr>

## **To Highlight 👍👌**

Hello team, Great job you've done in this project and for the requested changes earlier. I liked the way you managed and displayed the events system. However, there are more mandatory changes that need to be done before it gets approved!
Please read the following comments:-

## **To Improve🙏 [REQUIRED 🛑]**

- [ ] I see that you're using Devise gem for authentication while in this project you should create your own authentication function that makes the user sign up with just the email
Screenshot

- [x] Following the previous comment, you should add integration tests for the authentication system that you've build.
integration_tests_requirement

- [x] Views should not have logic code inside, please refactor your views logic code inside the application_helper.rb file
- [x] Please update your routes.rb file to be just using the needed routes, not all the resources, i.e. resources :events, only: [:new, :create, :show, :index]
- [x] Please update your readme file screenshot to your current app design and improve the description of the project, also add instructions to run the project and tests i.e. bundle install, rails db:create and rails db:migrate and so on for all the needed CLIs to make the project run.

<hr>

## **Would be super nice to have 😊[NOT REQUIRED ⚠]**

- [ ] To deploy the app on a live demo link i.e. Heroku.
- [ ] To enhance the design to be more friendly, it'd be a great app for your portfolio.
- [ ] In the next projects try to make your commits more descriptive i.e. "add styles + changes" is not enough for a commit.

<hr>

And that's all, please submit a re-review once you're done with the required changes.
Good luck and happy coding 👨‍💻💻

**Comment by AbdelrhmanAmin** [comment]
Couldn't create authentication without devise.

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello @AbdelrhmanAmin**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [x] when redirecting to the event index after creating an event I get this error `ndefined method `edit_event_path' for #<ActionView::Base:0x0000000000c3a0>`. please make sure to fix this error. it that the edit_event_path does not exist.
- [x] in the events and user controller you should have a variable for previous and upcoming events like this `@past_events = Event.previous_events` and remove it from the views. this helps to increase the performance of the app.
- [x] please make sure to add indexes for the creator_id, user_id, and the event_id columns.
- [x] please make sure to do all of the changes to the readme file asked by the previous tse.


---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @AbdelrhmanAmin  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implemented the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [x] There is an error running migration; with the following error: `SQLite3::SQLException: duplicate column name: creator_id`
=> Be careful while adding indexes, actually adding index doesn't require you to re-add a field which were already present in your table
- [x] So you are still required to implement this requested change from the prvious reviewer `please make sure to add indexes for the creator_id, user_id, and the event_id columns.` and make sure you don't mess up with the old migrations
- [x] You have a bug in event editing, actually when I click on `edit event` button, I want to get that event's data already pre-populated in the fields of the form, currently your project is not behaving like that, look at the screenshot below, and fix that as well, as a hint, that issue has something to do with `app/views/events/_form.html.erb` partial, the way you structured it, you need to work on it again!
Screenshot from 2021-01-20 08-13-21

___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Review by javitocor — APPROVED** [review]
# STATUS APPROVED 🥇 

Congratulations  🥳  you have met all the requirements 👏 
You can move into the next project 🚀 
Good luck with it! 😺

**Pull request merged**